src/reducers/versions.tsx

```diff
@@ -1069,12 +1072,13 @@ export const fetchDiff = ({
       dispatch(actions.abortFetchVersion({ versionId: headVersionId }));
       dispatch(errorsActions.addError({ error: response.error }));
     } else {
-      if (!getVersionInfo(versionsState, response.id)) {
+      if (forceReloadVersion || !getVersionInfo(versionsState, response.id)) {
         dispatch(
           actions.loadVersionInfo({
             version: response,
           }),
         );
+        dispatch(fileTreeActions.unsetForVersionId());
```

Comment:
I see what you're doing here but this is fragile because `fetchDiff()` should not be the one responsible to decide that the `fileTree` state should be reset. In fact, Redux makes it easy for `fileTree` itself to decide and I think that's a better approach:

<details>
<summary>Patch</summary>

```diff
diff --git a/src/reducers/fileTree.tsx b/src/reducers/fileTree.tsx
index 8b5d0dd..527d3b0 100644
--- a/src/reducers/fileTree.tsx
+++ b/src/reducers/fileTree.tsx
@@ -3,7 +3,12 @@ import log from 'loglevel';
 import { Reducer } from 'redux';
 import { ActionType, deprecated, getType } from 'typesafe-actions';
 
-import { EntryStatusMap, Version, viewVersionFile } from './versions';
+import {
+  actions as versionsActions,
+  EntryStatusMap,
+  Version,
+  viewVersionFile,
+} from './versions';
 import { ThunkActionCreator } from '../configureStore';
 import {
   createAdjustedQueryString,
@@ -499,10 +504,10 @@ export const getTree = (
   return treeState.tree;
 };
 
-const reducer: Reducer<FileTreeState, ActionType<typeof actions>> = (
-  state = initialState,
-  action,
-): FileTreeState => {
+const reducer: Reducer<
+  FileTreeState,
+  ActionType<typeof actions | typeof versionsActions>
+> = (state = initialState, action): FileTreeState => {
   switch (action.type) {
     case getType(actions.buildTree): {
       const { comparedToVersionId, version } = action.payload;
@@ -515,10 +520,17 @@ const reducer: Reducer<FileTreeState, ActionType<typeof actions>> = (
         tree,
       };
     }
-    case getType(actions.unsetForVersionId): {
+    case getType(versionsActions.loadVersionInfo): {
+      let forVersionId = state.forVersionId;
+      const { version } = action.payload;
+      if (forVersionId === version.id) {
+        // When *re-loading* a version, invalidate the current tree. This
+        // will force it to be rebuilt using the reloaded version object.
+        forVersionId = undefined;
+      }
       return {
         ...state,
-        forVersionId: undefined,
+        forVersionId,
       };
     }
     default:
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index 58fc74c..b8f788e 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -1078,7 +1078,6 @@ export const fetchDiff = ({
             version: response,
           }),
         );
-        dispatch(fileTreeActions.unsetForVersionId());
       }
       if (
         !getEntryStatusMap({

```

</details>

Reply:
Great. I did not know we can do this.